src/lib/Syntax.hs

```diff
@@ -895,6 +932,10 @@ instance HasVars Expr where
     Hof e   -> foldMap freeVars e
     Case e alts resultTy ->
       freeVars e <> freeVars alts <> freeVars resultTy
+    RecordCons    items a -> freeVars items <> freeVars a
+    RecordSplit   items a -> freeVars items <> freeVars a
+    VariantLift   items a -> freeVars items <> freeVars a
+    VariantSplit  items a -> freeVars items <> freeVars a
```

Comment:
What does `VariantSplit` do? I understand that you can split off a prefix of a record, but how can you split a prefix from a variant? That feels very much like a partial function, so shouldn't it be done only in case patterns?

Reply:
It's actually a total function. Just as splitting a record gives you (effectively) a product type of two records, splitting a variant gives you (effectively) a sum type of two variants. In other words, we have something like
```
-- partially applied because there's not really a type for a finite prefix of fields
VariantSplit {foo:Int | bar:Float}
    : {foo:Int | bar:Float | ...rest} -> ({foo:Int | bar:Float} | {|...rest})
```
We need it to handle pattern-matching on variants whose type we don't yet know. For instance,
```
case v of
  {|foo=x|} -> f x
  {|bar=y|} -> g y
  {|foo|bar|...other|} -> h other
```
is problematic because we might not yet know the concrete type of `v` (it might have a polymorphic tail) and so we don't yet know what the list of alternatives to emit in the core IR is. So during inference, this desugars to
```
case VariantSplit {foo:?0 | bar:?1} v of
  Left prefix -> case prefix of
                   {|foo=x|} -> f x
                   {|bar=y|} -> g y
  Right other -> h other
```
After splitting, we know exactly what fields can appear in the `Left` branch, so we can emit them in the right order.

(In practice, this is only actually produced when checking `case` statements; there's no user-level syntax for `VariantSplit`.)

Reply:
Ok, great. Thanks for the explanation. Can you please add some comments to the code (something short + maybe a link to this comment?), because that's quite a non-trivial construct.